cli/src/extension_api.ts

```diff
@@ -0,0 +1,176 @@
+// TODO: Figure out how to make API async-safe.
+// TODO: Always use String in API, since &str can't easily cross JS boundary.
+// TODO: Auto-generate interface types?
+// TODO: Auto-generate the entire file?
+// TODO: Document methods.
```

Comment:
These are the remaining things that should be considered before merging this PR, the main points are the following:

Currently our API isn't really async safe, as soon as one async block is held across another, a mutable borrow error will show up. Of course it's possible to await everything immediately, but that kinda defeats the purpose of making it async. I believe we'll have to reconsider our PhylumApi initialization approach (ping @andreaphylum).

Some of our APIs take a `&str`. It might be possible to share string references between JavaScript and Rust, but even if it is that doesn't sound like a great idea and would likely be very convoluted. So I'd recommend defining APIs with only types that can be easily shared (I've fixed all existing APIs in this PR).

Lastly, this file probably shouldn't have to be written by hand. I've attempted to write out all the interface types by hand but quickly gave up after considering the maintenance overhead. Generating this entire file with a proc macro should be possible if we play our hands right, but at the very least I'd like to generate the interface types by hand.

That said, the interface types aren't strictly necessary, so technically we could also just point users to some API documentation where these types are publicly documented.

Reply:
I think immutably shared `&str` may not be a concern. I assume those strings are owned by v8, and IIRC strings in JS are an immutable primitive, so we have a guarantee that a `&str` is not mutated across async boundaries.

Our current implementation of API initialization is more complex than I'd like, although I do have a few ideas about how to get rid of the "lazy initialized-once future" abstraction. This still doesn't solve the fundamental problem of holding guards across async boundaries. Even if we were to hold a mutex guard from an immutable `Ref` borrow, we can still trigger a deadlock in as simple a way as

```typescript
await Promise.all([Deno.core.opAsync('first_api'), Deno.core.opAsync('second_api')])
```

As soon as I have some spare cycles, I want to do a deep dive in `deno_core`'s code to understand how its concurrency model interacts with v8, and the invariants it upholds.

Reply:
> I think immutably shared &str may not be a concern. I assume those strings are owned by v8, and IIRC strings in JS are an immutable primitive, so we have a guarantee that a &str is not mutated across async boundaries.

The concern is that `&str` does not exist as a type in JavaScript. All you have is `string` and that is not the same.

> As soon as I have some spare cycles, I want to do a deep dive in deno_core's code to understand how its concurrency model interacts with v8, and the invariants it upholds.

I'm not sure how this is related to `v8`'s concurrency model? Fundamentally this issue is all Rust, it's all about being able to call two API functions without resolving the first one before calling the second.

Reply:
> The concern is that &str does not exist as a type in JavaScript. All you have is string and that is not the same.

We should investigate what exactly this distinction is. Per my (limited) understanding, there is a serialization layer between v8 and Rust, and data gets in/comes out of Deno extensions via this layer. If Deno is able to zero-copy serialize a v8-owned JS `string` into a `&str` reference (would be sufficient to create a slice out of the length and the underlying pointer to the utf-8 string, if it's available...), we do not need to create owned values.

> I'm not sure how this is related to v8's concurrency model? Fundamentally this issue is all Rust, it's all about being able to call two API functions without resolving the first one before calling the second.

Having a `PhylumApi` object behind an immutable reference would solve this at the root; all of its methods currently bind to `&self`, and we may have as many immutable `Ref`s as we want. I'll open an issue and work on a fix for this later.

Reply:
For prosperity from DMs:

The issue is here is specifically that Deno does not accept using a `&str` here and doing so will result in an error. While deno might be able to handle a `&str`, it will always deserialize the JavaScript `string` as `String` and does not do any implicit conversions:

```
[2022-06-20T17:56:31Z ERROR phylum] Execution failed: TypeError: Error parsing args at position 1: invalid type: string "f16e99b2-cfdc-4293-ac6a-1bb040b594f1", expected a borrowed string
        at Object.opAsync (deno:core/01_core.js:152:35)
        at Function.getJobStatus (deno:phylum:18:32)
        at file:///home/chris/.local/share/phylum/extensions/test/main.ts:6:35
```